### ISSUE TITLE
HTTP session returned status 429

### ISSUE THREAD

**Issue by t3lmoazevedo** [BJReplay/ha-solcast-solar#199]
The integration 4.2.3 stop working with the following error:

Logger: custom_components.solcast_solar.solcastapi
Source: custom_components/solcast_solar/solcastapi.py:475
integration: Solcast PV Forecast ([documentation](https://github.com/BJReplay/ha-solcast-solar), [issues](https://github.com/BJReplay/ha-solcast-solar/issues))
First occurred: 9:08:05 AM (27 occurrences)
Last logged: 9:11:06 AM

**HTTP session returned status 429/Try again later in __sites_data(), trying cache**

![image](https://github.com/user-attachments/assets/0d46c6c9-88a8-457c-a046-950f5380574d)

**Comment by BJReplay** [comment]
Hi there @t3lmoazevedo I see you've updated your issue while I'm writing...

I had written: 

_Version 4.0.26 is a **very old version**_

But I see you've changed to 4.2.23.

Anway, I had already written:

The Solcast Servers have been throwing lots of 429 errors recently, and there's not much we can do about that.

It looks like you've hit them during setup.

Can you confirm that you got this error message when you were first installing, or was it working, and then it stopped?

As always, debug logs (as described in #38 are helpful.

I'm not sure what you see when you click on the three vertical dots on the error message - if it allows you to retry or refresh, then that's the spot to retry.  The error message suggests that it will retry, and it might do that (I don't think we've actually run into a complete failure like this.

Otherwise, try adding a service with the Add Service link below.

It will attempt to take you through the setup, and hopefully it will work.

**Comment by t3lmoazevedo** [comment]
> It looks like you've hit them during setup.
> 
> Can you confirm that you got this error message when you were first installing, or was it working, and then it stopped?

Hi BJReplay, thanks for the feedback.
Yes I lost values yesterday, and reinstalled everything today and keeps getting this error even when I do Configure on the 3 dots:

![image](https://github.com/user-attachments/assets/7e9d056c-344b-44a4-85bf-aab63ca3653b)

It's a matter of wait for the servers get a different reply? Or is anything else that we can do to bypass this?

**Comment by autoSteve** [comment]
Before you edited @t3lmoazevedo I was going to say, _"Wow. That is truly a version from the archives... v4.2.3 currently."_

History? Have you just updated from v4.0.26 and are now getting the same issue?

This tells me that your `/config/solcast-sites.json` does not exist, or that you switched Solcast accounts entirely, so the integration does not trust the contents of `solcast-sites.json`, so it must connect to Solcast. Which has failed with a, "Go away! We're busy!"

If you have just updated from v4.0.26 then that was released before I had had even contributed the sites cache, so the cache will definitely not exist of that scenario is the case. Ergo, Solcast is busy, integration cannot continue without sites because we can't make stuff up. It's not us, it's them, etc.

(I have been getting 429s for the past two and half hours. Frustrating when you're trying to develop things...)

**Comment by autoSteve** [comment]
> Config flow could not be loaded

That's a new one, but I've seen it before. Refresh browser page.

**Comment by autoSteve** [comment]
Actually, BJ, the initial load only does three retries to get sites, with a five second pause in between each. If there is a sites cache then it gives up immediately on 429 and loads it instead.

I believe there is no sites cache in this case because upgrading from ancient.

**Comment by t3lmoazevedo** [comment]
> This tells me that your `/config/solcast-sites.json` does not exist,

I can confirm that after reinstaling everything I dont have that json file on my /config folder anymore.
Is it possible to create it manually? Can you send me the text content for it?
Thanks

**Comment by autoSteve** [comment]
_From which version have you upgraded from???_

**Comment by t3lmoazevedo** [comment]
> From which version have you upgraded from???

I was still using a old 4.0.26 version (probably not from this same repository), and deleted everything to fresh install this lasted one.

**Comment by t3lmoazevedo** [comment]
Ok, seems its really a matter of waiting retries :)
After a few ones, it got initialized! Great to have values again.
Thanks you guys again the helpful support ;)

![image](https://github.com/user-attachments/assets/1c6d320c-5bc0-4441-94de-853e33b6bec3)

![image](https://github.com/user-attachments/assets/cf8a58f7-c59f-424b-b53a-434dd8be53b0)

**Comment by BJReplay** [comment]
> Thanks you guys again the helpful support ;)

A pleasure!

**Comment by autoSteve** [comment]
So the world has shifted steadily along. A history session, but welcome to this repo!

Back in the v4.0.26 days Oziee committed GitHub suicide because he could not stand any more users reporting 429/"we're busy" status issues, then getting all uppity with him as if it was his fault.

@BJReplay grabbed the torch and resurrected a viable version of the integration from other forks.

@autoSteve joined the fray as a collab. around 4.0.31 IIRC, and straight away realised that there was a serious issue with hitting 429 status. So I did something about it.

I built in a cache, called `solcast-solar.json`, which contained the response from Solcast of "tell me about my sites" API call. The content will be specific to _your_ sites. I cannot tell you what is in it, because I don't know. Anyway, on 429 at startup the first thing the integration does is try the cache.

Given your latest post and closure of this issue you now have this file. Check it out! It is a godsend.

I also built in tenacious retries of forecast fetch a while ago, which also was not a thing. The ten retries (with back-off interval) are what BJ was referring to.

Long story short: Welcome again. And enjoy a far better Solcast integration experience.

Check out the integration options. You will be blown away by the progress, noted in the release notes in readme.

**Comment by autoSteve** [comment]
(Oh, and let's not forget an _auto-update_ feature introduced recently, wich may mean you could do away with a separate automation to fetch forecast updates...)

**Comment by t3lmoazevedo** [comment]
> not forget an _auto-update_ feature introduced recently,

> not forget an _auto-update_ feature introduced recently,

I'm using tha old automation to upade the 10 requests during the whole day.
Is it better or worst from the new auto-update feature?

`alias: "Energia: Solcast update"
description: ""
mode: single
triggers:
  - value_template: >-
      {% set nr = as_datetime(state_attr('sun.sun','next_rising')) | as_local %}
      {% set ns = as_datetime(state_attr('sun.sun','next_setting')) | as_local
      %} {% set api_request_limit = 10 %} {% if nr > ns %}
        {% set nr = nr - timedelta(hours = 24) %} 
      {% endif %} {% set hours_difference = (ns - nr) %} {% set interval_hours =
      hours_difference / api_request_limit %} {% set ns = namespace(match =
      false) %} {% for i in range(api_request_limit) %}
        {% set start_time = nr + (i * interval_hours) %}
        {% if ((start_time - timedelta(seconds=30)) <= now()) and (now() <= (start_time + timedelta(seconds=30))) %}
          {% set ns.match = true %}
        {% endif %}
      {% endfor %} {{ ns.match }}
    trigger: template
conditions:
  - condition: sun
    before: sunset
    after: sunrise
actions:
  - data: {}
    action: solcast_solar.update_forecasts`

> Long story short: Welcome again. And enjoy a far better Solcast integration experience.

Once again, you guys rock! And thanks for that ;)

**Comment by gcoan** [comment]
> I'm using tha old automation to upade the 10 requests during the whole day.
> Is it better or worst from the new auto-update feature?

There's probably not much if anything in it, using the automation within Solcast integration is one less automation for you to maintain, or keep it separate if you want to have your own control.
Swings and roundabouts.

And yes, latest version does indeed rock !

**Comment by autoSteve** [comment]
Auto-update from sunrise to sunset would do a very similar job.

Disable your automation, head into the integration options and enable auto-update sunrise/sunset, and enjoy less moving parts in HA.
